TrackletAlgorithm/TrackMerger.cc

```diff
@@ -0,0 +1,176 @@
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_3,
+  BXType bx_o,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::TrackWord> &trackWord_o,
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_0_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_1_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_2_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::BarrelStubWord> &barrelStubWords_3_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_0_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_1_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_2_o, 
+  hls::stream<TrackFit<NBarrelStub, NDiskStub>::DiskStubWord> &diskStubWords_3_o 
+  )
+{
+  #pragma HLS inline
+  static ComparisonModule comparisonModule[kNComparisonModules];
+  #pragma HLS array_partition variable=comparisonModule complete dim=0
+
+  // TrackStruct unmergedTracks[kMaxTrack];
+  // #pragma HLS array_partition variable=unmergedTracks complete dim=0
+  LOOP_Input:
+  for (unsigned int i = 0; i < kMaxTrack; i++){ 
+    #pragma HLS pipeline II=1
+    TrackStruct tracks[kNBuffers]; 
+    #pragma HLS array_partition variable=tracks complete dim=0
+    loadTrack(trackWord,
+              barrelStubWords_0,
+              barrelStubWords_1,
+              barrelStubWords_2,
+              barrelStubWords_3,
+              diskStubWords_0,
+              diskStubWords_1,
+              diskStubWords_2,
+              diskStubWords_3,
+              tracks[0]);
+    LOOP_ProcTracks:
+    for (unsigned int j = 0; j < kNComparisonModules; j++){
+      #pragma HLS unroll
+      comparisonModule[j].process(tracks[j],tracks[j+1]);
+
+    }
+    //putting straight into fifo in dataflow region
+    // fillUnmerged(tracks[kNComparisonModules],unmergedTracks, i);
+
+    // send unmerged tracks sent to output
+    unloadTrack(tracks[kNComparisonModules],
+                trackWord_o,
+                barrelStubWords_0_o,
+                barrelStubWords_1_o,
+                barrelStubWords_2_o,
+                barrelStubWords_3_o,
+                diskStubWords_0_o,
+                diskStubWords_1_o,
+                diskStubWords_2_o,
+                diskStubWords_3_o);
+
+  }
+
+  LOOP_OutputCM:
+  for (unsigned int nModule = 0; nModule < kNComparisonModules; nModule++){
+    #pragma HLS unroll
+    TrackStruct& masterTk = comparisonModule[nModule].getMasterTrackStruct();
+    unloadTrack(masterTk,
+                trackWord_o,
+                barrelStubWords_0_o,
+                barrelStubWords_1_o,
+                barrelStubWords_2_o,
+                barrelStubWords_3_o,
+                diskStubWords_0_o,
+                diskStubWords_1_o,
+                diskStubWords_2_o,
+                diskStubWords_3_o);
+
+  }
```

Comment:
Can you explain what this loop is doing?

Reply:
This loop is sending out the master track of each ComparisonModule after it has finished processing tracks.

Reply:
OK, I think I get what it's doing now. Thanks.